general/communication.hpp

```diff
@@ -26,25 +26,67 @@
 namespace mfem
 {
 
-/** @brief A simple convenience class that calls MPI_Init() at construction and
+/** @brief A simple singleton class that calls MPI_Init() at construction and
     MPI_Finalize() at destruction. It also provides easy access to
     MPI_COMM_WORLD's rank and size. */
+class Mpi
+{
+public:
+   /// Singleton creation with Mpi::Init();
+   static void Init() { Init_(NULL, NULL); }
+   /// Singleton creation with Mpi::Init(argc,argv);
+   static void Init(int &argc, char **&argv) { Init_(&argc, &argv); }
+   /// Access to the singleton with Mpi::Session()
+   static Mpi &Session()
+   {
+      Mpi &mpi = Instance();
+      MFEM_VERIFY(mpi.initialized, "MPI not initialized!");
+      return mpi;
+   }
+   static bool IsInitialized() { return Instance().initialized; }
+   int WorldRank() const { return world_rank; }
+   int WorldSize() const { return world_size; }
+   bool Root() const { return world_rank == 0; }
+private:
+   /// Initialize MPI
+   static void Init_(int *argc, char ***argv)
+   {
+      Mpi &mpi = Instance();
+      MFEM_VERIFY(!mpi.initialized, "MPI already initialized!")
+      MPI_Init(argc, argv);
+      MPI_Comm_rank(MPI_COMM_WORLD, &mpi.world_rank);
+      MPI_Comm_size(MPI_COMM_WORLD, &mpi.world_size);
+      mpi.initialized = true;
+   }
+   /// Finalize MPI
+   ~Mpi() { if (initialized) { MPI_Finalize(); } }
+   /// Access the singleton instance
+   static Mpi &Instance()
+   {
+      static Mpi mpi;
+      return mpi;
+   }
+   /// Prevent direct construction of objects of this class
+   Mpi() { }
+   int world_rank = 0;
+   int world_size = 0;
+   bool initialized = false;
```

Comment:
This class may interact better with other code MPI code if we use `MPI_Initialized` and `MPI_Finalized` to check the MPI library status. We already use these calls in `mfem_error`.

Reply:
OK, I made some changes to do that